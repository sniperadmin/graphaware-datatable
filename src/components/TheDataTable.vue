<template>
  <v-container>
    <!-- SECTION: Parent table -->
    <v-data-table
      :headers="headersCopy"
      :items="dataCopy"
      :single-expand="singleExpand"
      :expanded.sync="expanded"
      data-test="parent"
      item-key="Identification number"
      show-expand
      calculate-widths
      dense
      class="elevation-1"
    >
      <template v-slot:item.actions="{ item }">
        <v-icon
          small
          @click="deleteItem(item)"
        >
          mdi-delete
        </v-icon>
      </template>
      <template v-slot:expanded-item="{ headers, item }">
        <td :colspan="headers.length">
          <template v-for="(relVal, relKey, relIndex) in item.kids">
            <v-subheader>
              {{ relKey }}
            </v-subheader>

            <!-- SECTION: Relatives -->
            <v-data-table
              :headers="computedHeaders(relVal.records).headers"
              :items="computedItems(relVal.records).map(m => m.data)"
              :single-expand="singleExpand"
              :expanded.sync="childExpanded"
              item-key="Relative ID"
              @click:row="(i, s) => s.expand(!s.isExpanded)"
              show-expand
              hide-default-footer
              class="elevation-1 mb-4"
            >
              <template v-slot:item.actions="{ item }">
                <v-icon
                  small
                  @click="deleteItem(item)"
                >
                  mdi-delete
                </v-icon>
              </template>
              <template v-slot:expanded-item="{ headers, item }">
                <td :colspan="headers.length">
                  <template v-for="(firstVal, firstKey, firstIndex) in item.kids">
                    <v-subheader>
                      {{ firstKey }}
                    </v-subheader>

                    <!-- SECTION: Phones -->
                    <v-data-table
                      :headers="computedHeaders(firstVal.records).headers"
                      :items="computedItems(firstVal.records).map(m => m.data)"
                      :single-expand="singleExpand"
                      :expanded.sync="firstExpanded"
                      item-key="Phone ID"
                      @click:row="(i, s) => s.expand(!s.isExpanded)"
                      show-expand
                      hide-default-footer
                      class="elevation-1 mb-4"
                    >
                      <!-- ... -->
                    </v-data-table>
                    <!-- SECTION: ./Phones -->
                  </template>
                </td>
              </template>
            </v-data-table>
            <!-- SECTION: ./Relatives -->
          </template>
        </td>
      </template>
    </v-data-table>
    <!-- SECTION: ./Parent table -->
  </v-container>
</template>

<script lang="ts">
  import Vue from 'vue'
  import { hydrateHeadersPerObject, hydrateItemsPerObject, deleteObject } from '@/utils/recursive-hydrators'

  export default Vue.extend({
    name: 'TheDataTable',
    props: {
      data: {
        type: Array,
        required: true,
      }
    },

    data() {
      return {
        expanded: [],
        childExpanded: [],
        firstExpanded: [],
        singleExpand: false,
        headersCopy: null,
        dataCopy: null
      }
    },

    methods: {
      initialize() {
        this.headersCopy = this.computedHeaders(this.data).headers
        this.dataCopy = this.computedItems(this.data).map(o => o.data)
      },
      deleteItem(item) {
        const res = deleteObject(this.dataCopy, item)
        console.log(res);
      },
      computedHeaders(data) {
        if (data) {
          return hydrateHeadersPerObject(data)[0]
        }
      },
      computedItems(data) {
        if (data) {
          return hydrateItemsPerObject(data)[0]
        }
      }
    },

    created() {
      this.initialize()
    },
  })
</script>
