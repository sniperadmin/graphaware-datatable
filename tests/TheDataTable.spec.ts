import { shallowMount, createLocalVue } from '@vue/test-utils'
import TheDataTable from '@/components/TheDataTable.vue'
import fakeData from "../src/utils/example-data.json"

let wrapper: any;

describe('TheDataTable', () => {
  beforeEach(() => {
    const localVue = createLocalVue()
    wrapper = shallowMount(TheDataTable, {
      localVue,
      propsData: {
        data: fakeData
      }
    })
  })

  it('should mount', () => {
    expect(wrapper.vm).toBeTruthy()
  });

  it('should display parent table', () => {
    const table = wrapper.find("[data-test='parent']")
    expect(table.exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
    console.log(table.props('headers'));
    expect(table.props('headers')).toEqual([
      {
        text: 'Identification number',
        value: 'Identification number',
        sortable: true
      },
      { text: 'Name', value: 'Name', sortable: true },
      { text: 'Gender', value: 'Gender', sortable: true },
      { text: 'Risk', value: 'Risk', sortable: true },
      { text: 'Hair length', value: 'Hair length', sortable: true },
      { text: 'IQ', value: 'IQ', sortable: true },
      { text: 'Admission date', value: 'Admission date', sortable: true },
      { text: 'Last breakdown', value: 'Last breakdown', sortable: true },
      { text: 'Yearly fee', value: 'Yearly fee', sortable: true },
      {
        text: 'Knows the Joker?',
        value: 'Knows the Joker?',
        sortable: true
      },
      { text: 'kids', value: 'kids', sortable: true, align: ' d-none' }
    ])
  })
})
