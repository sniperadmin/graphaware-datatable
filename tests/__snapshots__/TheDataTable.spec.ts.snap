// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheDataTable should display parent table 1`] = `
"<v-container-stub tag=\\"div\\">
  <v-data-table-stub mobilebreakpoint=\\"600\\" items=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" options=\\"[object Object]\\" sortby=\\"\\" sortdesc=\\"\\" customsort=\\"[Function]\\" page=\\"1\\" itemsperpage=\\"10\\" groupby=\\"\\" groupdesc=\\"\\" customgroup=\\"[Function]\\" locale=\\"en-US\\" customfilter=\\"[Function]\\" serveritemslength=\\"-1\\" itemkey=\\"Identification number\\" value=\\"\\" expanded=\\"\\" noresultstext=\\"$vuetify.dataIterator.noResultsText\\" nodatatext=\\"$vuetify.noDataText\\" loadingtext=\\"$vuetify.dataIterator.loadingText\\" selectablekey=\\"isSelectable\\" loaderheight=\\"4\\" headers=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" showexpand=\\"true\\" dense=\\"true\\" calculatewidths=\\"true\\" expandicon=\\"$expand\\" itemclass=\\"\\" data-test=\\"parent\\" class=\\"elevation-1\\"></v-data-table-stub>
</v-container-stub>"
`;
